nengo/builder/connection.py

```diff
@@ -213,14 +210,19 @@ def get_prepost_signal(is_pre):
 
     # Build learning rules
     if conn.learning_rule is not None:
-        model.add_op(PreserveValue(model.sig[conn]['weights']))
-
         rule = conn.learning_rule
-        if is_iterable(rule):
-            for r in itervalues(rule) if isinstance(rule, dict) else rule:
-                model.build(r)
-        elif rule is not None:
-            model.build(rule)
+        rule = [rule] if not is_iterable(rule) else rule
+        targets = []
+        for r in itervalues(rule) if isinstance(rule, dict) else rule:
+            model.build(r)
+            targets.append(r.modifies)
+
+        if 'encoders' in targets:
+            model.add_op(PreserveValue(model.sig[conn.post_obj]['encoders']))
+        if 'decoders' in targets or 'weights' in targets:
+            if weights.ndim < 2:
+                raise ValueError("'transform' must be a matrix for learning")
```

Comment:
Another minor nitpick: should we say "2-dimensional array" instead of matrix to avoid people accidentally giving us an `np.matrix` instance?